Add a Prometheus metric for HVAC action

knyar opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add a Prometheus metric that indicates current HVAC action (heating, cooling, idle, etc).

Example output:
```
# HELP climate_action HVAC action
# TYPE climate_action gauge
climate_action{action="off",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 0.0
climate_action{action="heating",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 0.0
climate_action{action="cooling",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 0.0
climate_action{action="drying",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 0.0
climate_action{action="idle",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 1.0
climate_action{action="fan",domain="climate",entity="climate.bedroom",friendly_name="Bedroom"} 0.0
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

knyar committed:
Add a Prometheus metric for current HVAC action

homeassistant commented:
Hi @knyar, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31945).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -287,6 +296,23 @@ def _handle_climate(self, state):
+            }
+            metric = self._metric(
+                "action",
+                self.prometheus_cli.Gauge,
```
This doesn't make any sense. You can't graph this on a gauge. We shouldn't publish this as a number metric.

knyar replied:
An alternative would be to export state as a metric label, with metric value being 0 or 1 depending on whether a given state is active. So for a given entity there will always be only one state with the value 1, and all other states with the value of 0. Like this:

```
# HELP current_action Current HVAC action
# TYPE current_action gauge
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="off"} 1.0
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="heating"} 0.0
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="cooling"} 0.0
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="drying"} 0.0
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="idle"} 0.0
current_action{domain="climate",entity="climate.bedroom",friendly_name="Bedroom",action="fan"} 0.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="off"} 0.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="heating"} 1.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="cooling"} 0.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="drying"} 0.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="idle"} 0.0
current_action{domain="climate",entity="climate.bathroom",friendly_name="Bathroom",action="fan"} 0.0
```

This is a more conventional way to do this in Prometheus, however will involve exporting 6 metrics per entity. This could be a bit confusing, since I suspect most entities do not actually support all 6 possible states (so we will be exporting a metric with a value of 0 that can never have a value of 1 for a given entity).

Would you prefer this over encoding the state in the value? Or is there a better alternative you have in mind?

balloob replied:
I don't know anything about Promotheus or what best practices would be. Just know that you can't convert an enum into an int 🙂

knyar replied:
Both approaches – encoding state in the value, or as a separate label – are common in Prometheus. For example, if you look at the `node_exporter` (Prometheus binary for exporting machine-level metrics) you would see the former used for [runit service state](https://github.com/prometheus/node_exporter/blob/2477c5c67dff7e7655a9d466450235e9c9eac193/collector/runit.go#L87) and the latter for [systemd service state](https://github.com/prometheus/node_exporter/blob/6ad94ae4bc5527037ce0e6b8462f58cb3b8ccef3/collector/systemd_linux.go#L254), for example.

Since you seem to be opposed to encoding state as a metric value, I have amended this change to use the label approach as described in my previous comment.

Could you please take another look?

knyar replied:
Hey @balloob, just wanted to ask if you are happier with the new revision and whether you have any other comments I can address. Thanks!

balloob replied:
I think that this should be reviewed by someone who understands Prometheus, which is not me.

knyar replied:
Could you please suggest someone specific?

knyar replied:
@balloob, ping? Could you please review this or nominate a reviewer?

balloob replied:
I don't know.

balloob commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -150,9 +154,10 @@ def handle_event(self, event):
-        if labels is None:
-            labels = ["entity", "friendly_name", "domain"]
+    def _metric(self, metric, factory, documentation, extra_labels=None):
+        labels = ["entity", "friendly_name", "domain"]
```
This seems like it's changing logic unrelated to this PR.

knyar replied:
I don't believe this function ever gets called with `extra_labels` set.

mweinelt replied:
They're passing `action` as an extra label, so it is related.

mweinelt replied:
It's removing the posibility to overwrite all labels, though I'm not sure that is being used anywhere.

knyar replied:
As I mentioned above, that argument was never used. I have checked that.

knyar commented:
What do I need to do to get this merged?

mweinelt commented:
The proposal in https://github.com/home-assistant/core/pull/31945/files#r380970673 is what is used for systemd unit state and it works pretty well. I would go that route.

I'm not happy about the metric name. Neither `action` nor `current_action` (current is just wrong in a metrics name) is properly namespaced. If it it is part of a climate state shouldn't it be `climate_action`?

In fact, why call it action at all? In `homeassistant/components/climate/const.py` it's called `HVAC_MODES`. So maybe `climate_mode`?

```python
# Only the fan is on, not fan and another mode like cool
HVAC_MODE_FAN_ONLY = "fan_only"

HVAC_MODES = [
    HVAC_MODE_OFF,
    HVAC_MODE_HEAT,
    HVAC_MODE_COOL,
    HVAC_MODE_HEAT_COOL,
    HVAC_MODE_AUTO,
    HVAC_MODE_DRY,
    HVAC_MODE_FAN_ONLY,
]
```

knyar commented:
Thanks for your comments, @mweinelt! I just wanted to confirm: are you one of the maintainers of this repo and would be able to approve and merge this PR?

> The proposal in https://github.com/home-assistant/core/pull/31945/files#r380970673 is what is used for systemd unit state and it works pretty well. I would go that route.

I believe that's exactly how it's implemented (following @balloob's comments from Feb). I've adjusted the PR description as well.

> I'm not happy about the metric name. Neither action nor current_action (current is just wrong in a metrics name) is properly namespaced. If it it is part of a climate state shouldn't it be climate_action?

I don't feel strongly about the name. I chose `current_action` to be consistent with other metric names (for example, `current_temperature_c`). I am not sure `climate` needs to be part of the metric name, since there are `domain` and `entity` labels that make it clear this is climate-specific (for example, `domain="climate",entity="climate.bedroom"`).

Would you be happy with the metric just being called `action`?

> In fact, why call it action at all? In homeassistant/components/climate/const.py it's called HVAC_MODES. So maybe climate_mode?

You would notice that "HVAC mode" (`ATTR_HVAC_MODE`) is different from "HVAC action" (`ATTR_HVAC_ACTION`). This metric is designed to track "HVAC action".

mweinelt commented:
> Thanks for your comments, @mweinelt! I just wanted to confirm: are you one of the maintainers of this repo and would be able to approve and merge this PR?
> 

No, I'm just another contributor trying to submit changes to the prometheus module. A maintainer asked me to review this commit. See https://github.com/home-assistant/core/pull/35216.

> > The proposal in https://github.com/home-assistant/core/pull/31945/files#r380970673 is what is used for systemd unit state and it works pretty well. I would go that route.
> 
> I believe that's exactly how it's implemented (following @balloob's comments from Feb). I've adjusted the PR description as well.
> 
> > I'm not happy about the metric name. Neither action nor current_action (current is just wrong in a metrics name) is properly namespaced. If it it is part of a climate state shouldn't it be climate_action?
> 
> I don't feel strongly about the name. I chose `current_action` to be consistent with other metric names (for example, `current_temperature_c`).

Then the other metrics should be fixed up as well, because `current` is a function that can be applied to a metric. The metric has history, it can't be `current`.

> I am not sure `climate` needs to be part of the metric name, since there are `domain` and `entity` labels that make it clear this is climate-specific (for example, `domain="climate",entity="climate.bedroom"`).
> 

I've skimmed the prometheus component and the climate handler is the only obvious oddball not namespacing the metric and using the `current_` prefix. This should be fixed.

> Would you be happy with the metric just being called `action`?
> 
> > In fact, why call it action at all? In homeassistant/components/climate/const.py it's called HVAC_MODES. So maybe climate_mode?
> 
> You would notice that "HVAC mode" (`ATTR_HVAC_MODE`) is different from "HVAC action" (`ATTR_HVAC_ACTION`). This metric is designed to track "HVAC action".

Wouldn't action just be the intended state, while mode represents the actual state?

You seem to have pushed unrelated commits into this branch now.

MartinHjelmare commented:
`hvac_mode` is what the user has set the climate device to, eg auto, heat, cool, while `hvac_action` is what the device is doing at any one time, eg heating, idle, cooling.

knyar commented:
> Then the other metrics should be fixed up as well, because current is a function that can be applied to a metric. The metric has history, it can't be current.

Sounds good. I am happy to work with you on fixing existing metric names, but I don't think it should be part of this PR.

> I know this might be redundant wrt labels, but those labels have a low cardinality anyway.

I've renamed the metric to `climate_action`.

> Wouldn't action just be the intended state, while mode represents the actual state?

Reading https://github.com/home-assistant/architecture/issues/22, I believe it's the other way around: 'mode' represents configuration (user intent), while 'action' is what's happening right now.

> You seem to have pushed unrelated commits into this branch now.

Oh sorry, I rebased this against `master` instead of `dev` by mistake. Fixed now.

mweinelt commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -289,6 +294,19 @@ def _handle_climate(self, state):
+            )
+            for action in CURRENT_HVAC_ACTIONS:
+                metric.labels(**dict(self._labels(state), action=action)).set(
+                    1 if action == current_action else 0
```
Not sure if that's prettier, but it encapsulates the semantics pretty well IMO:
```suggestion
                    int(action == current_action)
```

knyar replied:
Done

mweinelt commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -289,6 +294,19 @@ def _handle_climate(self, state):
+            metric = self._metric(
+                "climate_action",
+                self.prometheus_cli.Gauge,
+                "Current HVAC action",
```
The "Current" here should be removed as well.

balloob replied:
Correct, drop the "current". We should follow the naming of the attributes in Home Assistant, which is hvac_action.

knyar replied:
Removed "Current" from metric description.

mweinelt reviewed (approved):
LGTM. Can you still check why the CI checks are failing?

knyar commented:
> LGTM. Can you still check why the CI checks are failing?

One of the failures was caused by a formatting check, and I've adjusted the code to conform.

`azure-pipelines` tests are still failing, but they [seem to be doing that consistently](https://dev.azure.com/home-assistant/Core/_pipeline/analytics/stageawareoutcome?definitionId=75), so I don't think the failures are related to this change.

MartinHjelmare commented:
The plex test is unrelated to this PR and has been fixed on the dev branch.

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
@knyar as I mentioned to @mweinelt it would be good if the prometheus integration contributors can band together and discuss the current state of the integration and what changes you would like to do and document this in an architecture issue. The core team can help advice on home assistant best practices if there are questions about that.

Another way of helping moving this integration forward is to become a codeowner. We welcome contributors that volunteer for this role. Codeowners will get mentioned on issues tagged with the integration label and on PRs for the integration. It's good if there are multiple codeowners per integration to spread the workload and the codeowners can then review each other PRs too.

knyar commented:
Thanks for merging this, @MartinHjelmare.

I am happy to be one of the codeowners. I don't have any other perssing changes I'd like to make, but I've started a discussion about plans in https://github.com/home-assistant/architecture/issues/391

MartinHjelmare commented:
Please make a PR adding yourself as codeowner to the manifest.json file. Run `python3 -m script.hassfest` to update the codeowners file.
